dali/kernels/imgproc/convolution/convolution_cpu.h

```diff
@@ -0,0 +1,362 @@
+   * at construction. The result is stored in pixel stored at `accum`.
+   */
+  template <typename W>
+  void CalculateDot(W* __restrict__ accum, const W* __restrict__ window) const {
+    assert(elements_ == length_);
+    for (int c = 0; c < NumLanes(); c++) {
+      accum[c] = 0;
+    }
+    int window_idx = 0;
+    for (int buf_idx = start_; buf_idx < length_; buf_idx++, window_idx++) {
+      for (int c = 0; c < NumLanes(); c++) {
+        accum[c] += window[window_idx] * data_[buf_idx * NumLanes() + c];
+      }
+    }
+    for (int buf_idx = 0; buf_idx < end_; buf_idx++, window_idx++) {
+      for (int c = 0; c < NumLanes(); c++) {
+        accum[c] += window[window_idx] * data_[buf_idx * NumLanes() + c];
+      }
+    }
+  }
+
+  template <typename U, typename W>
+  void CalculateDot(U* __restrict__ output, const W* __restrict__ window, W scale) const {
+    std::array<W, max_lanes> tmp;
+    CalculateDot(tmp.data(), window);
+    for (int c = 0; c < NumLanes(); c++) {
+      output[c] = ConvertSat<U>(tmp[c] * scale);
+    }
+  }
+
+  int Size() const {
+    return elements_;
+  }
+
+  bool Empty() const {
+    return elements_ == 0;
+  }
+
+ private:
+  void WrapPosition(int& pos) const {
+    if (pos == length_) {
+      pos = 0;
+    }
+  }
+
+  int NumLanes() const {
+    return std::min(num_lanes_, max_lanes);
+  }
+
+  T* __restrict__ data_ = nullptr;
+  int start_ = 0;
+  int end_ = 0;  ///< next empty element
+  int elements_ = 0;
+  int length_ = 0;
+  int num_lanes_ = 0;
+};
+
+template <typename T, int max_lanes>
+void load_pixel_with_border(CyclicWindowWrapper<T, max_lanes>& cww, const T* in_ptr, int in_idx,
+                            int stride, int axis_size) {
+  cww.PushElement(in_ptr + boundary::idx_reflect_101(in_idx, axis_size) * stride);
+}
+
+template <typename T, int max_lanes>
+void reload_pixel_with_border(CyclicWindowWrapper<T, max_lanes>& cww, const T* in_ptr, int in_idx,
+                              int out_idx, int stride, int axis_size, int radius) {
+  // out_idx is currently the center of the window
+  // we won't look further than radius elements, so we will pick something from first half
+  // of the cyclic window
+  auto in_idx_clamped = boundary::idx_reflect_101(in_idx, axis_size);
+  auto window_start_idx = out_idx - radius;
+  auto in_window_idx = in_idx_clamped - window_start_idx;
+  cww.PushElement(cww.GetElementOffset(in_window_idx));
+}
+
+template <typename T, int max_lanes>
+void load_pixel_no_border(CyclicWindowWrapper<T, max_lanes>& cww, const T* in_ptr, int in_idx,
+                          int stride) {
+  cww.PushElement(in_ptr + in_idx * stride);
+}
+
+template <bool has_channels, typename Out, typename In, typename W, int ndim>
+void ConvolveInnerDim(Out* out, const In* in, const W* window, int window_size, int radius,
+                      const TensorShape<ndim>& shape, const TensorShape<ndim>& strides, W scale) {
```

Comment:
What is the difference between radius and window_size?

Reply:
diameter = window_size
radius = (diameter - 1) / 2

Reply:
So do you need as the function argument?